File: src/main/java/com/snackgame/server/member/domain/Member.java
package com.snackgame.server.member.domain;

import javax.persistence.DiscriminatorColumn;
import javax.persistence.Embedded;
import javax.persistence.Entity;
import javax.persistence.EntityListeners;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Inheritance;
import javax.persistence.ManyToOne;

import com.snackgame.server.common.domain.BaseEntity;

import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

@EntityListeners(MemberEntityListener.class)
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Getter
@Inheritance
@DiscriminatorColumn(name = "type")
@Entity
public class Member extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    @Embedded
    private Name name;
    @ManyToOne
    private Group group;
    @Embedded
    private ProfileImage profileImage = ProfileImage.EMPTY;
    @Embedded
    private final Status status = new Status();

    private boolean isValid = true;

    public Member(Name name) {
        this.name = name;
    }

    public Member(Name name, ProfileImage profileImage) {
        this.name = name;
        this.profileImage = profileImage;
    }

    public Member(Long id, Name name, Group group) {
        this.id = id;
        this.name = name;
        this.group = group;
    }

    public void changeNameTo(Name name) {
        this.name = name;
    }

    public void changeGroupTo(Group group) {
        this.group = group;
    }

    public void changeProfileImageTo(ProfileImage profileImage) {
        this.profileImage = profileImage;
    }

    public void invalidate() {
        this.isValid = false;
    }

    public String getNameAsString() {
        return name.getString();
    }

    public AccountType getAccountType() {
        return AccountType.SELF;
    }
}

File: src/main/java/com/snackgame/server/member/domain/Member.kt
package com.snackgame.server.member.domain

import com.snackgame.server.common.domain.BaseEntity
import org.hibernate.annotations.SQLDelete
import org.hibernate.annotations.Where
import javax.persistence.DiscriminatorColumn
import javax.persistence.Embedded
import javax.persistence.Entity
import javax.persistence.EntityListeners
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import javax.persistence.Id
import javax.persistence.Inheritance
import javax.persistence.ManyToOne

@EntityListeners(MemberEntityListener::class)
@SQLDelete(sql = "update member set is_valid = false where id = ?")
@Where(clause = "is_valid = true")
@Inheritance
@DiscriminatorColumn(name = "type")
@Entity
open class Member(
    name: Name,
    group: Group? = null,
    profileImage: ProfileImage = ProfileImage.EMPTY,
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long = 0
) : BaseEntity() {

    @Embedded
    var name: Name = name
        private set

    @ManyToOne
    var group: Group? = group
        private set

    @Embedded
    var profileImage: ProfileImage = profileImage
        private set

    @Embedded
    val status: Status = Status()

    private var isValid = true

    open fun changeNameTo(name: Name) {
        this.name = name
    }

    open fun changeGroupTo(group: Group?) {
        this.group = group
    }

    fun changeProfileImageTo(profileImage: ProfileImage) {
        this.profileImage = profileImage
    }

    fun getNameAsString(): String = name.string

    open val accountType: AccountType
        get() = AccountType.SELF
}
